Cannot generate schema for sealed trait with enum implementors

I'm trying to use `zio-schema` with `zio-http` to de/serialize request/response `Body`s. I have a sealed trait representing the response body of my API, and two implementors of that trait, both of which are enums, representing success bodies (with a `"data"` field) and failure bodies (with a `"message"` field), but cannot seem to generate a schema for this structure. The below scala-cli script is a minimal representation of what I've tried. This compiles but fails at runtime with a `ClassCastException`:

```scala
//> using dep dev.zio::zio:2.1.1
//> using dep dev.zio::zio-http:3.0.0-RC8

import zio.schema.{ DeriveSchema, Schema, TypeId }
import zio.schema.Schema.Sequence
import zio.schema.codec.BinaryCodec
import zio.schema.codec.JsonCodec.schemaBasedBinaryCodec
import zio.http.*

sealed trait HttpResponse[+D]

enum HttpSuccessResponse[+D](val data: D) extends HttpResponse[D]:
   case Created[E](entity: E) extends HttpSuccessResponse[E](entity)
   case Found[E](entity: E) extends HttpSuccessResponse[E](entity)
   case Updated[E](entity: E) extends HttpSuccessResponse[E](entity)
   case Deleted[E](entity: E) extends HttpSuccessResponse[E](entity)
   case Succeeded[E](entity: E) extends HttpSuccessResponse[E](entity)

enum HttpFailureResponse(val message: String) extends HttpResponse[String]:
   case ServerError(msg: String) extends HttpFailureResponse(msg)
   case RequestError(msg: String) extends HttpFailureResponse(msg)

given HttpSuccessResponseSchema[D](using Schema[D]): Schema[HttpSuccessResponse[D]] = DeriveSchema.gen

given HttpFailureResponseSchema: Schema[HttpFailureResponse] = DeriveSchema.gen

given HttpResponseSchema[D](using Schema[D]): Schema[HttpResponse[D]] = DeriveSchema.gen

val response = HttpSuccessResponse.Created("user")

val body = Body.from(response)

@main def run() =
   println(body.asString)
```

The full error is
```bash
Exception in thread "main" java.lang.ExceptionInInitializerError
        at run.main(error_repro.scala:34)
Caused by: java.lang.ClassCastException: class zio.schema.Schema$CaseClass1$$anon$7 cannot be cast to class zio.schema.Schema$CaseClass0 (zio.schema.Schema$CaseClass1$$anon$7 and zio.schema.Schema$CaseClass0 are in unnamed module of loader 'app')
        at zio.schema.codec.JsonCodec$JsonEncoder$.$anonfun$2(JsonCodec.scala:388)
        at zio.Chunk$Arr.mapChunk(Chunk.scala:1769)
        at zio.ChunkLike.map(ChunkLike.scala:121)
        at zio.ChunkLike.map$(ChunkLike.scala:39)
        at zio.Chunk.map(Chunk.scala:42)
        at zio.schema.codec.JsonCodec$JsonEncoder$.enumEncoder(JsonCodec.scala:389)
        at zio.schema.codec.JsonCodec$JsonEncoder$.schemaEncoder(JsonCodec.scala:205)
        at zio.schema.codec.JsonCodec$JsonEncoder$.encode(JsonCodec.scala:177)
        at zio.schema.codec.JsonCodec$$anon$2.encode(JsonCodec.scala:93)
        at zio.schema.codec.JsonCodec$$anon$2.encode(JsonCodec.scala:92)
        at zio.http.Body$.from(Body.scala:209)
        at error_repro$package$.<clinit>(error_repro.scala:13)
```

Am I just doing this incorrectly or is this something not supported?

--- a/zio-schema-derivation/shared/src/main/scala-3/zio/schema/DeriveSchema.scala
+++ b/zio-schema-derivation/shared/src/main/scala-3/zio/schema/DeriveSchema.scala
@@ -360,7 +360,9 @@ private case class DeriveSchema()(using val ctx: Quotes) {
     val cases = typesAndLabels.map { case (tpe, label) => deriveCase[T](tpe, label, newStack) }
 
     val numParentFields: Int = TypeRepr.of[T].typeSymbol.declaredFields.length
-    val isSimpleEnum: Boolean = !TypeRepr.of[T].typeSymbol.children.map(_.declaredFields.length).exists( _ > numParentFields )
+    val childrenFields = TypeRepr.of[T].typeSymbol.children.map(_.declaredFields.length)
+    val childrenFieldsConstructor = TypeRepr.of[T].typeSymbol.children.map(_.caseFields.length)
+    val isSimpleEnum: Boolean = childrenFieldsConstructor.forall( _ == 0) && childrenFields.forall( _ <= numParentFields)
     val hasSimpleEnumAnn: Boolean = TypeRepr.of[T].typeSymbol.hasAnnotation(TypeRepr.of[_root_.zio.schema.annotation.simpleEnum].typeSymbol)
 
     val docAnnotationExpr = TypeRepr.of[T].typeSymbol.docstring.map { docstring =>

--- a/zio-schema-derivation/shared/src/test/scala-3/zio/schema/VersionSpecificDeriveSchemaSpec.scala
+++ b/zio-schema-derivation/shared/src/test/scala-3/zio/schema/VersionSpecificDeriveSchemaSpec.scala
@@ -48,6 +48,23 @@ trait VersionSpecificDeriveSchemaSpec extends ZIOSpecDefault {
     @description("Red")
     case Red
 
+  enum NonSimpleEnum1:
+    case A(a: Int)
+
+  enum NonSimpleEnum2(a: Int):
+    case A(b: Int) extends NonSimpleEnum2(0)
+
+
+  enum NonSimpleEnum3(a: Int):
+    case A(b: Int) extends NonSimpleEnum3(b)
+
+  enum NonSimpleEnum4(val a: Int):
+    case A(override val a: Int) extends NonSimpleEnum4(a)
+
+  enum NonSimpleEnum5(a: Int, b: String):
+    case A extends NonSimpleEnum5(0, "")
+    case B(n: Int) extends NonSimpleEnum5(n, "")
+
   def versionSpecificSuite = Spec.labeled(
     "Scala 3 specific tests",
     suite("Derivation")(
@@ -67,6 +84,18 @@ trait VersionSpecificDeriveSchemaSpec extends ZIOSpecDefault {
       test("correctly assigns simpleEnum to enum") {
         val derived: Schema[Colour] = DeriveSchema.gen[Colour]
         assertTrue(derived.annotations == Chunk(simpleEnum(true)))
+      },
+      test("doesn't assigns simpleEnum to non-simple enum") {
+        val derived1: Schema[NonSimpleEnum1] = DeriveSchema.gen[NonSimpleEnum1]
+        val derived2: Schema[NonSimpleEnum2] = DeriveSchema.gen[NonSimpleEnum2]
+        val derived3: Schema[NonSimpleEnum3] = DeriveSchema.gen[NonSimpleEnum3]
+        val derived4: Schema[NonSimpleEnum4] = DeriveSchema.gen[NonSimpleEnum4]
+        val derived5: Schema[NonSimpleEnum5] = DeriveSchema.gen[NonSimpleEnum5]
+        assertTrue(derived1.annotations.isEmpty) &&
+        assertTrue(derived2.annotations.isEmpty) &&
+        assertTrue(derived3.annotations.isEmpty) &&
+        assertTrue(derived4.annotations.isEmpty) &&
+        assertTrue(derived5.annotations.isEmpty)
       },
       test("derive different annotations for parent and child in enum") {
         val parent = DeriveSchema.gen[ColourAnnotations]

